File: LeetCode/124. Binary Tree Maximum Path Sum/Solution.kt
package problem124

import shouldBeEqualTo

class Solution {
  private var max = Int.MIN_VALUE

  fun maxPathSum(root: TreeNode?): Int {
    // 1. define sub-problem
    // f(node) -> use node as edge, the MPS
    // result = max(f(allNodes...))
    //
    // 2. recursion formula
    // - leftTree + node;
    // - node + rightTree;
    // f(node) -> node.value + max(f(node.left), 0) + max(f(node.right), 0)
    // max = maxOf(max, leftTree + node + rightTree)

    f(root)
    return max
  }

  fun f(node: TreeNode?): Int {
    if (node == null) return 0

    val left = f(node.left).coerceAtLeast(0)
    val right = f(node.right).coerceAtLeast(0)

    max = maxOf(node.`val` + left + right, max)

    return maxOf(node.`val` + left, node.`val` + right)
  }
}

data class TreeNode(var `val`: Int) {
  var left: TreeNode? = null
  var right: TreeNode? = null
}

fun main() {
  val root = TreeNode(5).apply {
    left = TreeNode(4).apply {
      left = TreeNode(11).apply {
        left = TreeNode(7)
        right = TreeNode(2)
      }
    }
    right = TreeNode(8).apply {
      left = TreeNode(13)
      right = TreeNode(4).apply {
        right = TreeNode(1)
      }
    }
  }

  Solution().maxPathSum(root) shouldBeEqualTo 48


}

File: LeetCode/124. Binary Tree Maximum Path Sum/Solution.cc
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

#include <algorithm>  // for std::min
#include <limits>

class Solution {
  int max = std::numeric_limits<int>::min();

 public:
  int maxPathSum(TreeNode *root) {
    f(root);
    return max;
  }

 private:
  int f(TreeNode *node) {
    if (!node) return 0;
    auto left = std::max(f(node->left), 0);
    auto right = std::max(f(node->right), 0);
    max = std::max(max, left + node->val + right);
    return std::max(left + node->val, right + node->val);
  }
};

// test

#include <iostream>

void test(const std::vector<int> &height, int expected) {
  int ret = Solution().trap(height);
  std::cout << ret << " == " << expected << std::endl;
}

int main() {
  test(std::vector<int>{4, 2, 0, 3, 2, 5}, 9);
  test(std::vector<int>{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}, 6);
  test(std::vector<int>{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}, 6);
}
